src/tyr/route_serializer_osrm.cc

```diff
@@ -1598,10 +1547,273 @@ json::ArrayPtr serialize_legs(const google::protobuf::RepeatedPtrField<valhalla:
+    }
+
+    return cache[route_idx][leg_idx][n];
+  }
+};
+
+std::vector<std::vector<std::string>>
+summarize_route_legs(const google::protobuf::RepeatedPtrField<DirectionsRoute>& routes) {
+
+  route_summary_cache rscache(routes);
+
+  // vector 1: routes
+  // vector 2: legs
+  // string: unique summary for the route/leg
+  std::vector<std::vector<std::string>> all_summaries;
+  all_summaries.reserve(routes.size());
+
+  // Find the simplest summary for every leg of every route. Important note:
+  // each route should have the same number of legs. Hence, we only need to make
+  // unique the same leg (leg_idx) between all routes.
+  for (size_t route_i = 0; route_i < routes.size(); route_i++) {
+
+    size_t num_legs_i = routes[route_i].legs_size();
+    std::vector<std::string> leg_summaries;
+    leg_summaries.reserve(num_legs_i);
+
+    for (size_t leg_idx = 0; leg_idx < num_legs_i; leg_idx++) {
+
+      // we desire each summary to be comprised of at least two named segments.
+      // however, if only one is available that's all we can use.
+      size_t num_named_segments_i = rscache.num_named_segments_for_route_leg(route_i, leg_idx);
+      size_t num_named_segs_needed = std::min(MAX_USED_SEGMENTS, num_named_segments_i);
+
+      // Compare every jth route/leg summary vs the current ith route/leg summary.
+      // We desire to compute num_named_segs_needed, which is the number of named
+      // segments needed to uniquely identify the ith's summary.
+      for (size_t route_j = 0; route_j < routes.size(); route_j++) {
+
+        // avoid self
+        if (route_i == route_j)
+          continue;
+
+        size_t num_legs_j = routes[route_j].legs_size();
+
+        // there should be the same number of legs in every route. however, some
+        // unit tests break this rule, so we cannot enable this assert.
+        // assert(num_legs_i == num_legs_j);
+        if (leg_idx >= num_legs_j)
+          continue;
+
+        size_t num_named_segments_j = rscache.num_named_segments_for_route_leg(route_j, leg_idx);
+
+        size_t num_comparable = std::min(num_named_segments_i, num_named_segments_j);
+
+        // k is the number of named segments in the summary. It keeps going
+        // up by 1 until route_i's summary is different than the route_j's.
+        size_t k = 1;
+        for (; (k < num_comparable); k++) {
+          const std::string& summary_i = rscache.get_n_segment_summary(route_i, leg_idx, k);
+          const std::string& summary_j = rscache.get_n_segment_summary(route_j, leg_idx, k);
+          if (summary_i != summary_j)
+            break;
+        }
+
+        if (k > num_named_segs_needed) {
+          num_named_segs_needed = k;
+        }
+      }
+
+      std::string leg_summary =
+          rscache.get_n_segment_summary(route_i, leg_idx, num_named_segs_needed);
+
+      leg_summaries.emplace_back(leg_summary);
+    }
+
+    all_summaries.emplace_back(leg_summaries);
+  }
+
+  return all_summaries;
```

Comment:
```
            |----B---|
-----A------N        |---C------D--
            |                |
         --E|                |
         |                   |
         |                   |
         |                   G
         |                   |
         --------F------------
```
I guess it's reasonable to generate summaries +- like this
1)"A, (any of BC), D"
2)"A, (any of EFG), D"

imho we should look at the nodes where routes fork(as this is what makes them different) and join back.
Not sure if it's possible to implement having current "routes-response", but just my view.

Anyway the current solution code gets rid of summary duplication so I'm okay to merge it

Reply:
You are correct.  The nice thing about appending the "next longest leg" into the summary is it is quite deterministic.  The (any of ...) approach could be made deterministic by some heuristic... say, by choosing the longest.  :)

I appreciate the thoughts, thanks.